### PR TITLE
Retrieve image text pairs

### PR DESCRIPTION
**Pull request by sahel-sh**
[3/3] Retrieve complement candidates when enabled in config. The complement candidate of an original retrieved candidate has a complement modality so that candidate and complement candidate are always image-text pairs.
The complement candidate for each original candidate is retrieved by using the original candidate as an interactive query.
The complement candidate is the top most relevant candidate with complement modality but different from the original query.

### CODE REVIEW

**Comment by lim142857** [comment]
Thank you for your contributions! I noticed that the InteractiveRetriever requires a pre-built candidate index file to function correctly. To assist users with this setup, could we consider adding a script, such as bash `run_interactive_retriever_pipeline.sh`, that demonstrates the entire pipeline? This script would cover embedding, indexing, and loading the index for the interactive retriever and retrieve demo queries. Additionally, incorporating a step-by-step guide in the README could greatly enhance the user experience.

File: src/common/interactive_retriever.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+class InteractiveRetriever:
+    def __init__(self, cand_index_path: str, candidates_path: str, config):
+        # Set up seed for reproducibility
+        seed = config.seed + dist_utils.get_rank()
+        set_seed(seed)
+
+        # MSCOCO's dataset id is hardcoded since the dataset id and query/candidate modalities determine the instruction part of the prompt.
+        # MSCOCO's dataset supports prompt instructions for both image->text and text->image query->candidate modalities.
+        self.dataset_id = 9
```
Comment (lim142857) [inline src/common/interactive_retriever.py:40]: Is the InteractiveRetriever specifically designed for the MSCOCO dataset? I observed that the `self.dataset_id `and `task_id` assignments appear to be hardcoded.

Comment (lim142857) [inline src/common/interactive_retriever.py:40]: Is the InteractiveRetriever specifically designed for the MSCOCO dataset? I observed that the `self.dataset_id `and `task_id` assignments appear to be hardcoded.

File: src/common/interactive_retriever.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+from utils import build_model_from_config, set_seed
+from data.preprocessing.utils import unhash_did
+
+
+class QueryModality(Enum):
+    TEXT = "text"
+    IMAGE = "image"
+
+
+class InteractiveRetriever:
```
Comment (lim142857) [inline src/common/interactive_retriever.py:32]: I noticed that the InteractiveRetriever requires a pre-built candidate index file to function correctly. To assist users with this setup, could we consider adding a script, such as `run_interactive_retriever_pipeline.sh`, that demonstrates the entire pipeline? This script would cover embedding, indexing, and loading the index for the interactive retriever and retrieve demo queries. Additionally, incorporating a step-by-step guide in the README could greatly enhance the user experience.

**Review by lim142857 — CHANGES_REQUESTED** [review]
1. For the InteractiveRetriever, it seems that hardcoding for MSCOCO is in place. I think we should make it into a tool applicable to all tasks and datasets.
2. We can add a **run_interactive_retriever_pipeline.sh**, that demonstrates the entire pipeline.
Please see the detailed comments for the review :)

File: src/common/interactive_retriever.py (hunk 1 above)
Comment (sahel-sh) [reply src/common/interactive_retriever.py]: I changed the interactiveRetriever to be generic, but the way it is currently integrated with the mbeir_retriever is for retrieving complement candidates to create image text pairs and mscoco is a dataset that supports both text->image and image->text queries. Now the embeir retriever sets the dataset to mscoco for this task.

**Comment by sahel-sh** [comment]
@lim142857 I applied the requested changes, PTAL

**Pull request merged**